src/Agent.Listener/Configuration/CredentialProvider.cs

```diff
@@ -64,80 +70,46 @@ public override VssCredentials GetVssCredentials(IHostContext context)
-            {
-                try
-                {
-                    if (PlatformUtil.RunningOnWindows)
-                    {
-                        Process.Start(new ProcessStartInfo() { FileName = codeResult.VerificationUrl, UseShellExecute = true });
-                    }
-                    else if (PlatformUtil.RunningOnLinux)
-                    {
-                        Process.Start(new ProcessStartInfo() { FileName = "xdg-open", Arguments = codeResult.VerificationUrl });
-                    }
-                    else
-                    {
-                        throw new NotImplementedException("Unexpected platform");
-                    }
-                }
-                catch (Exception ex)
-                {
-                    // not able to open browser, ex: xdg-open/open is not installed.
-                    trace.Error(ex);
-                    term.WriteLine($"Fail to open browser. {codeResult.Message}");
-                }
-            }
+            var app = PublicClientApplicationBuilder.Create(_clientId).Build();
 
-            AuthenticationResult authResult = ctx.AcquireTokenByDeviceCodeAsync(codeResult).GetAwaiter().GetResult();
-            ArgUtil.NotNull(authResult, nameof(authResult));
-            trace.Info($"receive AAD auth result with {authResult.AccessTokenType} token");
+            var authResult = AcquireATokenFromCacheOrDeviceCodeFlowAsync(context, app, new string[] { _userImpersonationScope }).GetAwaiter().GetResult();
 
-            var aadCred = new VssAadCredential(new VssAadToken(authResult.AccessTokenType, authResult.AccessToken));
+            var aadCred = new VssAadCredential(new VssAadToken(authResult.TokenType, authResult.AccessToken));
             VssCredentials creds = new VssCredentials(null, aadCred, CredentialPromptType.DoNotPrompt);
             trace.Info("cred created");
-
+            _cacheCreds = creds;
             return creds;
         }
-
         public override void EnsureCredential(IHostContext context, CommandSettings command, string serverUrl)
         {
             ArgUtil.NotNull(context, nameof(context));
             Tracing trace = context.GetTrace(nameof(AadDeviceCodeAccessToken));
             trace.Info(nameof(EnsureCredential));
             ArgUtil.NotNull(command, nameof(command));
             CredentialData.Data[Constants.Agent.CommandLine.Args.Url] = serverUrl;
-            CredentialData.Data[Constants.Agent.CommandLine.Flags.LaunchBrowser] = command.GetAutoLaunchBrowser().ToString();
+        }
+
+        public async Task<AuthenticationResult> AcquireATokenFromCacheOrDeviceCodeFlowAsync(IHostContext context, IPublicClientApplication app, IEnumerable<String> scopes)
+        {
+            AuthenticationResult result = null;
+            var accounts = await app.GetAccountsAsync().ConfigureAwait(false);
+
+            if (accounts.Any())
```

Comment:
1/ does it need to be public?
2/ most probably this func doesn't need to exist at all - you already cache the credentials in the GetVssCredentials & that func anyway keeps creating new instance of public client for every call (see my other comment on that topic) anyway.

We should either drop our home-grown cache (and fully rely on MSALs) or other way around.